File: Data Structures & Algorithms/redundant-connection/submission-1.java
class Solution {
    private Map<Integer, List<Integer>> graph;

    public int[] findRedundantConnection(int[][] edges) {
        graph = new HashMap<>();

        for (int i = 0; i <= edges.length; i++) {
            graph.put(i, new ArrayList<>());
        }

        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            graph.get(u).add(v);
            graph.get(v).add(u);

            Set<Integer> visited = new HashSet<>();

            if (!(dfs(u, -1, visited))) {
                return edge;
            }
        }

        return new int[0];

    }

    private boolean dfs(int node, int parent, Set<Integer> visited) {
        if (visited.contains(node)) {
            return false;
        }

        visited.add(node);

        for (int n : graph.get(node)) {
            if (n == parent) continue;

            if (!(dfs(n, node, visited))) {
                return false;
            }
        }

        return true;

    }
}

File: Data Structures & Algorithms/valid-tree/submission-1.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        tree = defaultdict(list)

        for edge in edges:
            tree[edge[0]].append(edge[1])
            tree[edge[1]].append(edge[0])
        
        visited = set()

        def dfs(node, par):
            if node in visited:
                return False
            
            visited.add(node)
            for nei in tree[node]:
                if nei == par:
                    continue
                if not dfs(nei, node):
                    return False

            return True
        
        return dfs(0, -1) and n == len(visited)
